# EDA: Vehículos

In [1]:
import pandas as pd
import pyarrow
import requests
import matplotlib.pyplot as plt
from matplotlib.colors import LinearSegmentedColormap
import seaborn as sns
import numpy as np

# Alternative Fuel Vehicles US.csv

Marca, modelo y características técnicas de consumo de automóbiles híbridos.

In [2]:
afv = pd.read_csv("DATA/Alternative Fuel Vehicles US.csv")

afv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 882 entries, 0 to 881
Data columns (total 22 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Category                            882 non-null    object 
 1   Model                               882 non-null    object 
 2   Model Year                          594 non-null    float64
 3   Manufacturer                        882 non-null    object 
 4   Fuel                                882 non-null    object 
 5   All-Electric Range                  222 non-null    float64
 6   PHEV Total Range                    43 non-null     float64
 7   Alternative Fuel Economy City       169 non-null    float64
 8   Alternative Fuel Economy Highway    169 non-null    float64
 9   Alternative Fuel Economy Combined   262 non-null    float64
 10  Conventional Fuel Economy City      444 non-null    float64
 11  Conventional Fuel Economy Highway   444 non-n

In [3]:
afv.columns

Index(['Category', 'Model', 'Model Year', 'Manufacturer', 'Fuel',
       'All-Electric Range', 'PHEV Total Range',
       'Alternative Fuel Economy City', 'Alternative Fuel Economy Highway',
       'Alternative Fuel Economy Combined', 'Conventional Fuel Economy City',
       'Conventional Fuel Economy Highway',
       'Conventional Fuel Economy Combined', 'Transmission Type',
       'Transmission Make', 'Engine Type', 'Engine Size',
       'Engine Cylinder Count', 'Number of Passengers',
       'Heavy-Duty Power System', 'Notes', 'Drivetrain'],
      dtype='object')

Categoría: Sedan

Modelo

Año

Fabricante

Combustible

Autonomía eléctrica

Autonomía de enchufe

Autonomía eléctrica en ciudad

Autonomía eléctrica en autopista

Autonomía eléctrica combinada

Autonomía combustible en ciudad

Autonomía combustible en autopista

Autonomía combustible combinada

Tipo de caja (manual, automática)

Fabricante de la caja

Tipo de motor

Tamaño del motor

Cantidad de cilindros del motor

Cantidad de pasajeros

Sistema de trabajo con fuerza

Notas

Tracción (delantera, trasera)

In [4]:
afv["Model Year"].unique()

array([2022., 2021.,   nan])

Tenemos modelos híbridos únicamente de los años 2021 y 2022.

In [5]:
afv["Fuel"].unique()

array(['Hybrid Electric', 'Plug-in Hybrid Electric', 'Electric',
       'Biodiesel (B20)', 'Ethanol (E85)', 'Hydrogen Fuel Cell',
       'CNG - Bi-fuel', 'CNG - Compressed Natural Gas',
       'Propane - Bi-fuel', 'Propane',
       'LNG - Liquified Natural Gas|CNG - Compressed Natural Gas',
       'CNG - Compressed Natural Gas|LNG - Liquified Natural Gas',
       'Propane|CNG - Compressed Natural Gas',
       'CNG - Compressed Natural Gas|Plug-in Hybrid Electric',
       'Diesel/Hybrid Electric',
       'Propane - Bi-fuel|CNG - Bi-fuel|Propane|CNG - Compressed Natural Gas|Ethanol (E85)',
       'Propane - Bi-fuel|Propane|CNG - Bi-fuel|Ethanol (E85)|CNG - Compressed Natural Gas',
       'CNG - Compressed Natural Gas|Propane - Bi-fuel|E85/Hybrid Electric|CNG - Bi-fuel|Propane',
       'CNG - Bi-fuel|Ethanol (E85)|CNG - Compressed Natural Gas|Propane|Propane - Bi-fuel',
       'Propane - Bi-fuel|Propane|CNG - Compressed Natural Gas|CNG - Bi-fuel',
       'CNG - Compressed Natural Gas|Prop

Podemos conservar los que su combustible alternativo es eléctrico.

In [6]:
afv["Transmission Make"].unique()

array([nan, 'Allison', 'Vorza', 'Allison 3000', 'BYD',
       'Allison, Voith, ZF', '.', 'BAE', 'Allison, Eaton', 'Eaton',
       'Dana', 'Voith, Allison, ZF', 'Global', 'GM, Allison',
       'MYD 6-speed automatic', 'Allison B400', 'Allison 2100, B300',
       'Allison B300', 'Allison B300, B400',
       'Allison PTS 2200 or Ford 5-speed OD',
       'Allison PTS 2200 or Ford Torque Shift 6-speed automatic',
       'Allison 2500 RDS', 'Aisin', 'PACCAR', 'Eaton, Allison',
       'Mack Powershift 2-speed ',
       'Allison, Eaton, mDRIVE\xa0HD/XHD, Maxitorque ES', 'Allision',
       'Siemens', 'Allison, BAE', 'ZF, Allison, Voith',
       'Allison, Eaton Manual', 'Eaton Automated, Allison, Eaton Manual'],
      dtype=object)

In [7]:
afv["Transmission Type"].unique()

array(['Auto', nan, 'Automatic', 'Automatic|Manual', 'Manual|Automatic'],
      dtype=object)

In [8]:
afv["Alternative Fuel Economy City"].unique()

array([ nan,  81.,  78.,  72.,  62.,  76.,  73.,  64., 100., 109.,  94.,
        79.,  86.,  82., 125., 131.,  11.,  10.,  14.,  12.,  98., 105.,
        97.,  90.,  88., 110., 108., 104.,  18.,  16., 127., 132.,  59.,
        65., 116., 136., 134., 123., 117., 126., 115., 130., 121.,  92.,
       119., 118., 114., 113.,  75.,  71.,  69.,  74., 138., 124., 102.,
       107., 103., 129., 140.,  67., 106.,  96.,  17.,  68., 145.,  80.,
       141., 150.,  85.])

In [9]:
afv["PHEV Total Range"].unique()

array([ nan, 410., 390., 430., 320., 290., 340., 400., 520., 330., 620.,
       440., 420., 470., 370., 560., 460., 480., 550., 300., 640., 600.,
       510., 530., 490., 500.])

### EDA detallado: sobre filas de interés

In [10]:
afv["Category"].unique()

array(['Sedan/Wagon', 'SUV', 'Pickup', 'Van', 'Passenger Van/Shuttle Bus',
       'Street Sweeper', 'Refuse', 'Vocational/Cab Chassis', 'Tractor',
       'School Bus', 'Transit Bus', 'Step Van',
       'TractorVocational/Cab Chassis', 'Vocational/Cab ChassisVan',
       'Vocational/Cab ChassisTractor'], dtype=object)

In [11]:
afvv = afv[(afv["Category"] == "Sedan/Wagon") | (afv["Category"] == "SUV") | (afv["Category"] == "Van")]
afvv.shape

(528, 22)

Nos quedamos solo con los vehículos de tipo Sedan/Wagon, SUV y Van que son los propicios para la ciudad.

528 modelos distintos.

In [12]:
afvvd = afvv.dropna(how="all")
afvvd

Category                      Model  Model Year  \
0    Sedan/Wagon                        NSX      2022.0   
1    Sedan/Wagon                        A3       2022.0   
2    Sedan/Wagon                 A3 quattro      2022.0   
3    Sedan/Wagon         A4 allroad quattro      2022.0   
4    Sedan/Wagon                 A4 quattro      2022.0   
..           ...                        ...         ...   
711          Van  Transit 250/350 Cargo Van         NaN   
738          Van                   Zevo 600         NaN   
741          Van           Savana 2500/3500         NaN   
742          Van           Savana 2500/3500         NaN   
755          Van              EV Star Cargo         NaN   

                 Manufacturer             Fuel  All-Electric Range  \
0                       Acura  Hybrid Electric                 NaN   
1                        Audi  Hybrid Electric                 NaN   
2                        Audi  Hybrid Electric                 NaN   
3                        Audi  Hybrid Electric                 NaN   
4                        Audi  Hybrid Electric                 NaN   
..                        ...              ...                 ...   
711                      Ford  Hybrid Electric                 NaN   
738             GM BrightDrop         Electric                 NaN   
741                       GMC    Ethanol (E85)                 NaN   
742                       GMC  Biodiesel (B20)                 NaN   
755  GreenPower Motor Company         Electric                 NaN   

     PHEV Total Range  Alternative Fuel Economy City  \
0                 NaN                            NaN   
1                 NaN                            NaN   
2                 NaN                            NaN   
3                 NaN                            NaN   
4                 NaN                            NaN   
..                ...                            ...   
711               NaN                            NaN   
738               NaN                            NaN   
741               NaN                            NaN   
742               NaN                            NaN   
755               NaN                            NaN   

     Alternative Fuel Economy Highway  Alternative Fuel Economy Combined  ...  \
0                                 NaN                                NaN  ...   
1                                 NaN                                NaN  ...   
2                                 NaN                                NaN  ...   
3                                 NaN                                NaN  ...   
4                                 NaN                                NaN  ...   
..                                ...                                ...  ...   
711                               NaN                                NaN  ...   
738                               NaN                                NaN  ...   
741                               NaN                                NaN  ...   
742                               NaN                                NaN  ...   
755                               NaN                                NaN  ...   

     Conventional Fuel Economy Combined  Transmission Type  Transmission Make  \
0                                  21.0               Auto                NaN   
1                                  32.0               Auto                NaN   
2                                  31.0               Auto                NaN   
3                                  26.0               Auto                NaN   
4                                  29.0               Auto                NaN   
..                                  ...                ...                ...   
711                                 NaN                NaN                NaN   
738                                 NaN                NaN                NaN   
741                                 NaN                NaN                NaN   
74

Quitamos las filas nulas, no se elimina ninguna.

In [13]:
afvvd["Alternative Fuel Economy City"].sort_values(ascending=False)

555    150.0
445    145.0
553    141.0
281    140.0
563    140.0
       ...  
711      NaN
738      NaN
741      NaN
742      NaN
755      NaN
Name: Alternative Fuel Economy City, Length: 528, dtype: float64

Ordenamos los modelos de acuerdo a la mayor autonomía eléctrica en la ciudad. La unidad es Wh/Km

In [14]:
afvvd_ordenado = afvvd.loc[afvvd["Alternative Fuel Economy City"].sort_values(ascending=False).index]
afvvd_ordenado

Category                            Model  Model Year  \
555  Sedan/Wagon  Model 3 Standard Range Plus RWD      2021.0   
445  Sedan/Wagon                   Ioniq Electric      2021.0   
553  Sedan/Wagon           Model 3 Long Range AWD      2021.0   
281          SUV                      Model Y RWD      2022.0   
563          SUV       Model Y Standard Range RWD      2021.0   
..           ...                              ...         ...   
711          Van        Transit 250/350 Cargo Van         NaN   
738          Van                         Zevo 600         NaN   
741          Van                 Savana 2500/3500         NaN   
742          Van                 Savana 2500/3500         NaN   
755          Van                    EV Star Cargo         NaN   

                 Manufacturer             Fuel  All-Electric Range  \
555                     Tesla         Electric               263.0   
445                   Hyundai         Electric               170.0   
553                     Tesla         Electric               353.0   
281                     Tesla         Electric               244.0   
563                     Tesla         Electric               244.0   
..                        ...              ...                 ...   
711                      Ford  Hybrid Electric                 NaN   
738             GM BrightDrop         Electric                 NaN   
741                       GMC    Ethanol (E85)                 NaN   
742                       GMC  Biodiesel (B20)                 NaN   
755  GreenPower Motor Company         Electric                 NaN   

     PHEV Total Range  Alternative Fuel Economy City  \
555               NaN                          150.0   
445               NaN                          145.0   
553               NaN                          141.0   
281               NaN                          140.0   
563               NaN                          140.0   
..                ...                            ...   
711               NaN                            NaN   
738               NaN                            NaN   
741               NaN                            NaN   
742               NaN                            NaN   
755               NaN                            NaN   

     Alternative Fuel Economy Highway  Alternative Fuel Economy Combined  ...  \
555                             133.0                              142.0  ...   
445                             121.0                              133.0  ...   
553                             127.0                              134.0  ...   
281                             119.0                              129.0  ...   
563                             119.0                              129.0  ...   
..                                ...                                ...  ...   
711                               NaN                                NaN  ...   
738                               NaN                                NaN  ...   
741                               NaN                                NaN  ...   
742                               NaN                                NaN  ...   
755                               NaN                                NaN  ...   

     Conventional Fuel Economy Combined  Transmission Type  Transmission Make  \
555                                 NaN               Auto                NaN   
445                                 NaN               Auto                NaN   
553                                 NaN               Auto                NaN   
281                                 NaN               Auto                NaN   
563                                 NaN               Auto                NaN   
..                                  ...                ...                ...   
711                                 NaN                NaN                NaN   
738                                 NaN                NaN                NaN   
741        

In [15]:
afvvd["Fuel"].unique()

array(['Hybrid Electric', 'Plug-in Hybrid Electric', 'Electric',
       'Biodiesel (B20)', 'Ethanol (E85)', 'Hydrogen Fuel Cell',
       'Propane - Bi-fuel', 'CNG - Compressed Natural Gas',
       'CNG - Bi-fuel', 'Propane'], dtype=object)

In [16]:
afvvd_ordenado_elec = afvvd_ordenado[(afvvd_ordenado["Fuel"] == "Hybrid Electric") | (afvvd_ordenado["Fuel"] == "Plug-in Hybrid Electric") | (afvvd_ordenado["Fuel"] == "Electric")]
afvvd_ordenado_elec

Category                            Model  Model Year  \
555  Sedan/Wagon  Model 3 Standard Range Plus RWD      2021.0   
445  Sedan/Wagon                   Ioniq Electric      2021.0   
553  Sedan/Wagon           Model 3 Long Range AWD      2021.0   
281          SUV                      Model Y RWD      2022.0   
563          SUV       Model Y Standard Range RWD      2021.0   
..           ...                              ...         ...   
689          Van              E-Transit Cargo Van         NaN   
710          Van        Transit 250/350 Cargo Van         NaN   
711          Van        Transit 250/350 Cargo Van         NaN   
738          Van                         Zevo 600         NaN   
755          Van                    EV Star Cargo         NaN   

                 Manufacturer             Fuel  All-Electric Range  \
555                     Tesla         Electric               263.0   
445                   Hyundai         Electric               170.0   
553                     Tesla         Electric               353.0   
281                     Tesla         Electric               244.0   
563                     Tesla         Electric               244.0   
..                        ...              ...                 ...   
689                      Ford         Electric                 NaN   
710                      Ford         Electric                 NaN   
711                      Ford  Hybrid Electric                 NaN   
738             GM BrightDrop         Electric                 NaN   
755  GreenPower Motor Company         Electric                 NaN   

     PHEV Total Range  Alternative Fuel Economy City  \
555               NaN                          150.0   
445               NaN                          145.0   
553               NaN                          141.0   
281               NaN                          140.0   
563               NaN                          140.0   
..                ...                            ...   
689               NaN                            NaN   
710               NaN                            NaN   
711               NaN                            NaN   
738               NaN                            NaN   
755               NaN                            NaN   

     Alternative Fuel Economy Highway  Alternative Fuel Economy Combined  ...  \
555                             133.0                              142.0  ...   
445                             121.0                              133.0  ...   
553                             127.0                              134.0  ...   
281                             119.0                              129.0  ...   
563                             119.0                              129.0  ...   
..                                ...                                ...  ...   
689                               NaN                                NaN  ...   
710                               NaN                                NaN  ...   
711                               NaN                                NaN  ...   
738                               NaN                                NaN  ...   
755                               NaN                                NaN  ...   

     Conventional Fuel Economy Combined  Transmission Type  Transmission Make  \
555                                 NaN               Auto                NaN   
445                                 NaN               Auto                NaN   
553                                 NaN               Auto                NaN   
281                                 NaN               Auto                NaN   
563                                 NaN               Auto                NaN   
..                                  ...                ...                ...   
689                                 NaN          Automatic                NaN   
710                                 NaN                NaN                NaN   
711        

Filtramos donde el combustible alternativo es eléctrico y no otros tipos de combustión como el alcohol.

In [17]:
afvvd_ordenado_elec.columns

Index(['Category', 'Model', 'Model Year', 'Manufacturer', 'Fuel',
       'All-Electric Range', 'PHEV Total Range',
       'Alternative Fuel Economy City', 'Alternative Fuel Economy Highway',
       'Alternative Fuel Economy Combined', 'Conventional Fuel Economy City',
       'Conventional Fuel Economy Highway',
       'Conventional Fuel Economy Combined', 'Transmission Type',
       'Transmission Make', 'Engine Type', 'Engine Size',
       'Engine Cylinder Count', 'Number of Passengers',
       'Heavy-Duty Power System', 'Notes', 'Drivetrain'],
      dtype='object')

In [18]:
afvvd_ordenado_elec_column = afvvd_ordenado_elec[['Category', 'Model', 'Model Year', 'Manufacturer', 'Fuel',
       'All-Electric Range', 'PHEV Total Range',
       'Alternative Fuel Economy City',
       'Alternative Fuel Economy Combined', 'Conventional Fuel Economy City',
       'Conventional Fuel Economy Combined']]

In [19]:
afvvd_ordenado_elec_column.info()

<class 'pandas.core.frame.DataFrame'>
Index: 481 entries, 555 to 755
Data columns (total 11 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Category                            481 non-null    object 
 1   Model                               481 non-null    object 
 2   Model Year                          476 non-null    float64
 3   Manufacturer                        481 non-null    object 
 4   Fuel                                481 non-null    object 
 5   All-Electric Range                  218 non-null    float64
 6   PHEV Total Range                    43 non-null     float64
 7   Alternative Fuel Economy City       125 non-null    float64
 8   Alternative Fuel Economy Combined   218 non-null    float64
 9   Conventional Fuel Economy City      351 non-null    float64
 10  Conventional Fuel Economy Combined  351 non-null    float64
dtypes: float64(7), object(4)
memory usage: 45.1+ KB


In [20]:
afvvd_ordenado_elec_column.reset_index(inplace=True)

In [21]:
afvvd_ordenado_elec_column.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 481 entries, 0 to 480
Data columns (total 12 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   index                               481 non-null    int64  
 1   Category                            481 non-null    object 
 2   Model                               481 non-null    object 
 3   Model Year                          476 non-null    float64
 4   Manufacturer                        481 non-null    object 
 5   Fuel                                481 non-null    object 
 6   All-Electric Range                  218 non-null    float64
 7   PHEV Total Range                    43 non-null     float64
 8   Alternative Fuel Economy City       125 non-null    float64
 9   Alternative Fuel Economy Combined   218 non-null    float64
 10  Conventional Fuel Economy City      351 non-null    float64
 11  Conventional Fuel Economy Combined  351 non-n

In [22]:
afvvd_ordenado_elec_column_sin_index = afvvd_ordenado_elec_column.drop(columns=['index'])

In [23]:
afvvd_ordenado_elec_column_sin_index

Category                            Model  Model Year  \
0    Sedan/Wagon  Model 3 Standard Range Plus RWD      2021.0   
1    Sedan/Wagon                   Ioniq Electric      2021.0   
2    Sedan/Wagon           Model 3 Long Range AWD      2021.0   
3            SUV                      Model Y RWD      2022.0   
4            SUV       Model Y Standard Range RWD      2021.0   
..           ...                              ...         ...   
476          Van              E-Transit Cargo Van         NaN   
477          Van        Transit 250/350 Cargo Van         NaN   
478          Van        Transit 250/350 Cargo Van         NaN   
479          Van                         Zevo 600         NaN   
480          Van                    EV Star Cargo         NaN   

                 Manufacturer             Fuel  All-Electric Range  \
0                       Tesla         Electric               263.0   
1                     Hyundai         Electric               170.0   
2                       Tesla         Electric               353.0   
3                       Tesla         Electric               244.0   
4                       Tesla         Electric               244.0   
..                        ...              ...                 ...   
476                      Ford         Electric                 NaN   
477                      Ford         Electric                 NaN   
478                      Ford  Hybrid Electric                 NaN   
479             GM BrightDrop         Electric                 NaN   
480  GreenPower Motor Company         Electric                 NaN   

     PHEV Total Range  Alternative Fuel Economy City  \
0                 NaN                          150.0   
1                 NaN                          145.0   
2                 NaN                          141.0   
3                 NaN                          140.0   
4                 NaN                          140.0   
..                ...                            ...   
476               NaN                            NaN   
477               NaN                            NaN   
478               NaN                            NaN   
479               NaN                            NaN   
480               NaN                            NaN   

     Alternative Fuel Economy Combined  Conventional Fuel Economy City  \
0                                142.0                             NaN   
1                                133.0                             NaN   
2                                134.0                             NaN   
3                                129.0                             NaN   
4                                129.0                             NaN   
..                                 ...                             ...   
476                                NaN                             NaN   
477                                NaN                             NaN   
478                                NaN                             NaN   
479                                NaN                             NaN   
480                                NaN                             NaN   

     Conventional Fuel Economy Combined  
0                                   NaN  
1                                   NaN  
2                                   NaN  
3                                   NaN  
4                                   NaN  
..                                  ...  
476                                 NaN  
477                                 NaN  
478                                 NaN  
479                                 NaN  
480                                 NaN  

[481 rows x 11 columns]

Reiniciamos el índice.

## Resumen de las columnas:

Categoría: Conservamos Sedan/Wagon, SUV y Van, que son los tipos más comunes de ciudad.

Modelo

Año

Fabricante:

Combustible: Conservamos Hybrid Electric, Plug-in Hybrid Electric y Electric, es decir todos los que tienen energía eléctrica.

Autonomía eléctrica

Autonomía de enchufe/de carga rápida: Representa la distancia que el vehículo recorre únicamente utilizando la energía eléctrica.

Autonomía eléctrica en ciudad: Se ordenó el df a partir de esta columna.

Autonomía eléctrica en autopista: Se eliminó

Autonomía eléctrica combinada

Autonomía combustible en ciudad

Autonomía combustible en autopista: Se eliminó

Autonomía combustible combinada

Tipo de caja (manual, automática): Se eliminó

Fabricante de la caja: Se eliminó

Tipo de motor: Se eliminó

Tamaño del motor: Se eliminó

Cantidad de cilindros del motor: Se eliminó

Cantidad de pasajeros: Se eliminó porque no contenía datos.

Sistema de trabajo con fuerza: Se eliminó

Notas: Se eliminó

Tracción (delantera, trasera): Se eliminó

# Electric and Alternative Fuel Charging Stations

In [24]:
stations = pd.read_csv("DATA/Electric and Alternative Fuel Charging Stations.csv", low_memory=False)

stations.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 70406 entries, 0 to 70405
Data columns (total 65 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   Fuel Type Code                    70406 non-null  object 
 1   Station Name                      70406 non-null  object 
 2   Street Address                    70405 non-null  object 
 3   Intersection Directions           4492 non-null   object 
 4   City                              70406 non-null  object 
 5   State                             70406 non-null  object 
 6   ZIP                               70405 non-null  object 
 7   Plus4                             0 non-null      float64
 8   Station Phone                     65204 non-null  object 
 9   Status Code                       70406 non-null  object 
 10  Expected Date                     1218 non-null   object 
 11  Groups With Access Code           70406 non-null  object 
 12  Acce

In [25]:
stations.columns

Index(['Fuel Type Code', 'Station Name', 'Street Address',
       'Intersection Directions', 'City', 'State', 'ZIP', 'Plus4',
       'Station Phone', 'Status Code', 'Expected Date',
       'Groups With Access Code', 'Access Days Time', 'Cards Accepted',
       'BD Blends', 'NG Fill Type Code', 'NG PSI', 'EV Level1 EVSE Num',
       'EV Level2 EVSE Num', 'EV DC Fast Count', 'EV Other Info', 'EV Network',
       'EV Network Web', 'Geocode Status', 'Latitude', 'Longitude',
       'Date Last Confirmed', 'ID', 'Updated At', 'Owner Type Code',
       'Federal Agency ID', 'Federal Agency Name', 'Open Date',
       'Hydrogen Status Link', 'NG Vehicle Class', 'LPG Primary',
       'E85 Blender Pump', 'EV Connector Types', 'Country',
       'Intersection Directions (French)', 'Access Days Time (French)',
       'BD Blends (French)', 'Groups With Access Code (French)',
       'Hydrogen Is Retail', 'Access Code', 'Access Detail Code',
       'Federal Agency Code', 'Facility Type', 'CNG Dispenser N

Código tipo de combustible

Nombre de la estación

Dirección

Intersección

Ciudad

Estado

ZIP Code

Plus4: Servicio de carga eléctrica rápido

Teléfono de la estación

Latitud

Longitud

Tipo de conectores

Tipo de combustible

Capacidad total de combustible

In [26]:
stations.head(3)

Fuel Type Code                                  Station Name  \
0            CNG          Spire - Montgomery Operations Center   
1            CNG                           PS Energy - Atlanta   
2            CNG  Metropolitan Atlanta Rapid Transit Authority   

        Street Address                            Intersection Directions  \
0     2951 Chestnut St                                                NaN   
1     340 Whitehall St  From I-7585 N, exit 91 to Central Ave, left on...   
2  2424 Piedmont Rd NE                                                NaN   

         City State    ZIP  Plus4 Station Phone Status Code  ...  \
0  Montgomery    AL  36107    NaN           NaN           E  ...   
1     Atlanta    GA  30303    NaN  770-350-3000           E  ...   
2     Atlanta    GA  30324    NaN           NaN           E  ...   

  EV Pricing (French) LPG Nozzle Types Hydrogen Pressures Hydrogen Standards  \
0                 NaN              NaN                NaN                NaN   
1                 NaN              NaN                NaN                NaN   
2                 NaN              NaN                NaN                NaN   

  CNG Fill Type Code CNG PSI CNG Vehicle Class  LNG Vehicle Class  \
0                  B    3600                MD                NaN   
1                  Q    3600                MD                NaN   
2                  Q    3000                LD                NaN   

   EV On-Site Renewable Source  Restricted Access  
0                          NaN                NaN  
1                          NaN              False  
2                          NaN                NaN  

[3 rows x 65 columns]

In [27]:
stations["Plus4"].unique()

array([nan])

Corroboramos que Plus4 tenga datos distintos a NaN pero está vacía.

In [28]:
stations.isna().sum()

Fuel Type Code                     0
Station Name                       0
Street Address                     1
Intersection Directions        65914
City                               0
                               ...  
CNG PSI                        68809
CNG Vehicle Class              68784
LNG Vehicle Class              70248
EV On-Site Renewable Source    70036
Restricted Access              16971
Length: 65, dtype: int64

Varias columnas que no son de nuestro interés tienen datos nulos.

In [29]:
stations["Restricted Access"].unique()

array([nan, False, True], dtype=object)

In [30]:
stat = stations[['Fuel Type Code', 'Station Name', 'City', 'State', 'Status Code', 'Latitude', 'Longitude']]
stat.head(3)

Fuel Type Code                                  Station Name        City  \
0            CNG          Spire - Montgomery Operations Center  Montgomery   
1            CNG                           PS Energy - Atlanta     Atlanta   
2            CNG  Metropolitan Atlanta Rapid Transit Authority     Atlanta   

  State Status Code   Latitude  Longitude  
0    AL           E  32.367916 -86.267021  
1    GA           E  33.745843 -84.398837  
2    GA           E  33.821911 -84.367461

Nos quedamos solo con las columnas de interés.

In [31]:
stat_boro = stat[(stat["State"] == "NY") & ((stat["City"] == "Brooklyn") | (stat["City"] == "Staten Island") | (stat["City"] == "Manhattan") | (stat["City"] == "Bronx") | (stat["City"] == "Queens") )]
stat_boro.head(3)

Fuel Type Code                               Station Name      City State  \
7             CNG  Clean Energy - Greenpoint - National Grid  Brooklyn    NY   
8             CNG                   Canarsie - National Grid  Brooklyn    NY   
11            CNG       Con Edison - Van Nest Service Center     Bronx    NY   

   Status Code   Latitude  Longitude  
7            E  40.718037 -73.932309  
8            E  40.645540 -73.918344  
11           E  40.844188 -73.861572

Filtramos únicamente las pertenecientes al estado de NY y sus distritos.

In [32]:
stat_boro["City"].unique()

array(['Brooklyn', 'Bronx', 'Queens', 'Staten Island', 'Manhattan'],
      dtype=object)

In [33]:
stat_boro

Fuel Type Code                               Station Name      City  \
7                CNG  Clean Energy - Greenpoint - National Grid  Brooklyn   
8                CNG                   Canarsie - National Grid  Brooklyn   
11               CNG       Con Edison - Van Nest Service Center     Bronx   
13               CNG  Con Edison - College Point Service Center    Queens   
1500             CNG                                  Bronx Zoo     Bronx   
...              ...                                        ...       ...   
69624           ELEC                   Rapid Park - 44 State St  Brooklyn   
69640           ELEC                               906 Union St  Brooklyn   
69993           ELEC                            Supercharge NYC  Brooklyn   
70050           ELEC                            35 Red Cross Pl  Brooklyn   
70288           ELEC                     40 N 4th St Apartments  Brooklyn   

      State Status Code   Latitude  Longitude  
7        NY           E  40.718037 -73.932309  
8        NY           E  40.645540 -73.918344  
11       NY           E  40.844188 -73.861572  
13       NY           E  40.769623 -73.843069  
1500     NY           E  40.850018 -73.878128  
...     ...         ...        ...        ...  
69624    NY           E  40.691800 -73.998290  
69640    NY           E  40.672830 -73.971750  
69993    NY           E  40.713604 -73.933099  
70050    NY           E  40.699487 -73.989516  
70288    NY           E  40.718391 -73.963339  

[200 rows x 7 columns]

In [34]:
stat_boro["Fuel Type Code"].unique()

array(['CNG', 'ELEC', 'E85', 'BD'], dtype=object)

# ElectricCarData_Clean

Catálogo de autos eléctricos. Autonomía, tipo de carga, asientos y precio

In [35]:
ecdc = pd.read_csv("DATA/ElectricCarData_Clean.csv")

ecdc

Brand                            Model  AccelSec  TopSpeed_KmH  \
0         Tesla     Model 3 Long Range Dual Motor       4.6           233   
1    Volkswagen                         ID.3 Pure      10.0           160   
2      Polestar                                 2       4.7           210   
3           BMW                              iX3        6.8           180   
4         Honda                                e        9.5           145   
..           ...                              ...       ...           ...   
98       Nissan                       Ariya 63kWh       7.5           160   
99         Audi     e-tron S Sportback 55 quattro       4.5           210   
100      Nissan               Ariya e-4ORCE 63kWh       5.9           200   
101      Nissan   Ariya e-4ORCE 87kWh Performance       5.1           200   
102       Byton                 M-Byte 95 kWh 2WD       7.5           190   

     Range_Km  Efficiency_WhKm FastCharge_KmH RapidCharge PowerTrain  \
0         450              161            940         Yes        AWD   
1         270              167            250         Yes        RWD   
2         400              181            620         Yes        AWD   
3         360              206            560         Yes        RWD   
4         170              168            190         Yes        RWD   
..        ...              ...            ...         ...        ...   
98        330              191            440         Yes        FWD   
99        335              258            540         Yes        AWD   
100       325              194            440         Yes        AWD   
101       375              232            450         Yes        AWD   
102       400              238            480         Yes        AWD   

       PlugType  BodyStyle Segment  Seats  PriceEuro  
0    Type 2 CCS      Sedan       D      5      55480  
1    Type 2 CCS  Hatchback       C      5      30000  
2    Type 2 CCS   Liftback       D      5      56440  
3    Type 2 CCS        SUV       D      5      68040  
4    Type 2 CCS  Hatchback       B      4      32997  
..          ...        ...     ...    ...        ...  
98   Type 2 CCS  Hatchback       C      5      45000  
99   Type 2 CCS        SUV       E      5      96050  
100  Type 2 CCS  Hatchback       C      5      50000  
101  Type 2 CCS  Hatchback       C      5      65000  
102  Type 2 CCS        SUV       E      5      62000  

[103 rows x 14 columns]

In [36]:
ecdc.columns

Index(['Brand', 'Model', 'AccelSec', 'TopSpeed_KmH', 'Range_Km',
       'Efficiency_WhKm', 'FastCharge_KmH', 'RapidCharge', 'PowerTrain',
       'PlugType', 'BodyStyle', 'Segment', 'Seats', 'PriceEuro'],
      dtype='object')

Marca

Modelo

Tiempo de aceleración entre 0 a 100km/h

Velocidad máxima

Rendimiento Watts/Km

Fastcharge

Carga rápida

Tracción

Tipo de enchufe

Cuerpo del auto

Segmento

Cantidad de asientos

Precio en euros

In [37]:
ecdc.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Data columns (total 14 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Brand            103 non-null    object 
 1   Model            103 non-null    object 
 2   AccelSec         103 non-null    float64
 3   TopSpeed_KmH     103 non-null    int64  
 4   Range_Km         103 non-null    int64  
 5   Efficiency_WhKm  103 non-null    int64  
 6   FastCharge_KmH   103 non-null    object 
 7   RapidCharge      103 non-null    object 
 8   PowerTrain       103 non-null    object 
 9   PlugType         103 non-null    object 
 10  BodyStyle        103 non-null    object 
 11  Segment          103 non-null    object 
 12  Seats            103 non-null    int64  
 13  PriceEuro        103 non-null    int64  
dtypes: float64(1), int64(5), object(8)
memory usage: 11.4+ KB


In [38]:
ecdc.columns

Index(['Brand', 'Model', 'AccelSec', 'TopSpeed_KmH', 'Range_Km',
       'Efficiency_WhKm', 'FastCharge_KmH', 'RapidCharge', 'PowerTrain',
       'PlugType', 'BodyStyle', 'Segment', 'Seats', 'PriceEuro'],
      dtype='object')

In [39]:
ecdc["Seats"].unique()

array([5, 4, 7, 6, 2])

In [40]:
ecdc1 = ecdc[['Brand', 'Model', 'Range_Km',
       'Efficiency_WhKm', 'FastCharge_KmH', 'RapidCharge', 'PowerTrain',
       'Seats', 'PriceEuro']]

ecdc1

Brand                            Model  Range_Km  Efficiency_WhKm  \
0         Tesla     Model 3 Long Range Dual Motor       450              161   
1    Volkswagen                         ID.3 Pure       270              167   
2      Polestar                                 2       400              181   
3           BMW                              iX3        360              206   
4         Honda                                e        170              168   
..           ...                              ...       ...              ...   
98       Nissan                       Ariya 63kWh       330              191   
99         Audi     e-tron S Sportback 55 quattro       335              258   
100      Nissan               Ariya e-4ORCE 63kWh       325              194   
101      Nissan   Ariya e-4ORCE 87kWh Performance       375              232   
102       Byton                 M-Byte 95 kWh 2WD       400              238   

    FastCharge_KmH RapidCharge PowerTrain  Seats  PriceEuro  
0              940         Yes        AWD      5      55480  
1              250         Yes        RWD      5      30000  
2              620         Yes        AWD      5      56440  
3              560         Yes        RWD      5      68040  
4              190         Yes        RWD      4      32997  
..             ...         ...        ...    ...        ...  
98             440         Yes        FWD      5      45000  
99             540         Yes        AWD      5      96050  
100            440         Yes        AWD      5      50000  
101            450         Yes        AWD      5      65000  
102            480         Yes        AWD      5      62000  

[103 rows x 9 columns]

Mantenemos las columnas de nuestro interés.

# ElectricCarData_Norm

Datos normalizados. En su mayoría son strings que permiten su legibilidad pero dificultan su análisis matemático

In [41]:
ecdn = pd.read_csv("DATA/ElectricCarData_Norm.csv")

ecdn

Brand                            Model     Accel  TopSpeed   Range  \
0         Tesla     Model 3 Long Range Dual Motor   4.6 sec  233 km/h  450 km   
1    Volkswagen                         ID.3 Pure  10.0 sec  160 km/h  270 km   
2      Polestar                                 2   4.7 sec  210 km/h  400 km   
3           BMW                              iX3    6.8 sec  180 km/h  360 km   
4         Honda                                e    9.5 sec  145 km/h  170 km   
..           ...                              ...       ...       ...     ...   
98       Nissan                       Ariya 63kWh   7.5 sec  160 km/h  330 km   
99         Audi     e-tron S Sportback 55 quattro   4.5 sec  210 km/h  335 km   
100      Nissan               Ariya e-4ORCE 63kWh   5.9 sec  200 km/h  325 km   
101      Nissan   Ariya e-4ORCE 87kWh Performance   5.1 sec  200 km/h  375 km   
102       Byton                 M-Byte 95 kWh 2WD   7.5 sec  190 km/h  400 km   

    Efficiency FastCharge              RapidCharge         PowerTrain  \
0    161 Wh/km   940 km/h  Rapid charging possible    All Wheel Drive   
1    167 Wh/km   250 km/h  Rapid charging possible   Rear Wheel Drive   
2    181 Wh/km   620 km/h  Rapid charging possible    All Wheel Drive   
3    206 Wh/km   560 km/h  Rapid charging possible   Rear Wheel Drive   
4    168 Wh/km   190 km/h  Rapid charging possible   Rear Wheel Drive   
..         ...        ...                      ...                ...   
98   191 Wh/km   440 km/h  Rapid charging possible  Front Wheel Drive   
99   258 Wh/km   540 km/h  Rapid charging possible    All Wheel Drive   
100  194 Wh/km   440 km/h  Rapid charging possible    All Wheel Drive   
101  232 Wh/km   450 km/h  Rapid charging possible    All Wheel Drive   
102  238 Wh/km   480 km/h  Rapid charging possible    All Wheel Drive   

       PlugType  BodyStyle Segment  Seats  PriceEuro  
0    Type 2 CCS      Sedan       D      5      55480  
1    Type 2 CCS  Hatchback       C      5      30000  
2    Type 2 CCS   Liftback       D      5      56440  
3    Type 2 CCS        SUV       D      5      68040  
4    Type 2 CCS  Hatchback       B      4      32997  
..          ...        ...     ...    ...        ...  
98   Type 2 CCS  Hatchback       C      5      45000  
99   Type 2 CCS        SUV       E      5      96050  
100  Type 2 CCS  Hatchback       C      5      50000  
101  Type 2 CCS  Hatchback       C      5      65000  
102  Type 2 CCS        SUV       E      5      62000  

[103 rows x 14 columns]

In [42]:
ecdn.columns

Index(['Brand', 'Model', 'Accel', 'TopSpeed', 'Range', 'Efficiency',
       'FastCharge', 'RapidCharge', 'PowerTrain', 'PlugType', 'BodyStyle',
       'Segment', 'Seats', 'PriceEuro'],
      dtype='object')

In [43]:
ecdn.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103 entries, 0 to 102
Data columns (total 14 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Brand        103 non-null    object
 1   Model        103 non-null    object
 2   Accel        103 non-null    object
 3   TopSpeed     103 non-null    object
 4   Range        103 non-null    object
 5   Efficiency   103 non-null    object
 6   FastCharge   103 non-null    object
 7   RapidCharge  103 non-null    object
 8   PowerTrain   103 non-null    object
 9   PlugType     103 non-null    object
 10  BodyStyle    103 non-null    object
 11  Segment      103 non-null    object
 12  Seats        103 non-null    int64 
 13  PriceEuro    103 non-null    int64 
dtypes: int64(2), object(12)
memory usage: 11.4+ KB


In [44]:
type(ecdn["TopSpeed"][0])

str

Esta tabla responde a la anterior y nos permite utilizarla como diccionario de datos.

# Light Duty Vehicles

Autos de "trabajo liviano" híbridos, autonomía eléctrica y de combustible, no incluye precio.

In [45]:
ldv = pd.read_csv("DATA/Light Duty Vehicles.csv")

ldv.head(3)

Vehicle ID  Fuel ID  Fuel Configuration ID  Manufacturer ID  Category ID  \
0       13044       45                    9.0              365           27   
1       12854       45                    9.0              377           27   
2       12842       45                    9.0              377           27   

        Model  Model Year  Alternative Fuel Economy City  \
0         NSX        2022                            NaN   
1         A3         2022                            NaN   
2  A3 quattro        2022                            NaN   

   Alternative Fuel Economy Highway  Alternative Fuel Economy Combined  ...  \
0                               NaN                                NaN  ...   
1                               NaN                                NaN  ...   
2                               NaN                                NaN  ...   

               Manufacturer URL     Category  Fuel Code             Fuel  \
0         http://www.acura.com/  Sedan/Wagon       HYBR  Hybrid Electric   
1  http://progress.audiusa.com/  Sedan/Wagon       HYBR  Hybrid Electric   
2  http://progress.audiusa.com/  Sedan/Wagon       HYBR  Hybrid Electric   

  Fuel Configuration Name Electric-Only Range  PHEV Total Range PHEV Type  \
0         Hybrid Electric                 NaN               NaN       NaN   
1         Hybrid Electric                 NaN               NaN       NaN   
2         Hybrid Electric                 NaN               NaN       NaN   

  Notes Drivetrain  
0   NaN        AWD  
1   NaN        FWD  
2   NaN        AWD  

[3 rows x 29 columns]

In [46]:
ldv.columns

Index(['Vehicle ID', 'Fuel ID', 'Fuel Configuration ID', 'Manufacturer ID',
       'Category ID', 'Model', 'Model Year', 'Alternative Fuel Economy City',
       'Alternative Fuel Economy Highway', 'Alternative Fuel Economy Combined',
       'Conventional Fuel Economy City', 'Conventional Fuel Economy Highway',
       'Conventional Fuel Economy Combined', 'Transmission Type',
       'Engine Type', 'Engine Size', 'Engine Cylinder Count',
       'Engine Description', 'Manufacturer', 'Manufacturer URL', 'Category',
       'Fuel Code', 'Fuel', 'Fuel Configuration Name', 'Electric-Only Range',
       'PHEV Total Range', 'PHEV Type', 'Notes', 'Drivetrain'],
      dtype='object')

In [58]:
ldv.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3008 entries, 0 to 3007
Data columns (total 29 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   Vehicle ID                          3008 non-null   int64  
 1   Fuel ID                             3008 non-null   int64  
 2   Fuel Configuration ID               2973 non-null   float64
 3   Manufacturer ID                     3008 non-null   int64  
 4   Category ID                         3008 non-null   int64  
 5   Model                               3008 non-null   object 
 6   Model Year                          3008 non-null   int64  
 7   Alternative Fuel Economy City       1595 non-null   float64
 8   Alternative Fuel Economy Highway    1505 non-null   float64
 9   Alternative Fuel Economy Combined   467 non-null    float64
 10  Conventional Fuel Economy City      2044 non-null   float64
 11  Conventional Fuel Economy Highway   2046 no

In [59]:
ldv.isna().sum()

Vehicle ID                               0
Fuel ID                                  0
Fuel Configuration ID                   35
Manufacturer ID                          0
Category ID                              0
Model                                    0
Model Year                               0
Alternative Fuel Economy City         1413
Alternative Fuel Economy Highway      1503
Alternative Fuel Economy Combined     2541
Conventional Fuel Economy City         964
Conventional Fuel Economy Highway      962
Conventional Fuel Economy Combined    2245
Transmission Type                       86
Engine Type                            797
Engine Size                            134
Engine Cylinder Count                  565
Engine Description                    1037
Manufacturer                             0
Manufacturer URL                       164
Category                                 0
Fuel Code                                0
Fuel                                     0
Fuel Config

Este dataset en particular podría no ser apropiado para nuestro análisis.

# taxi+_zone_lookup

Zonas y tipos de servicios de NYC

In [47]:
tzl = pd.read_csv("DATA/taxi+_zone_lookup.csv")

tzl.head(3)

LocationID Borough                     Zone service_zone
0           1     EWR           Newark Airport          EWR
1           2  Queens              Jamaica Bay    Boro Zone
2           3   Bronx  Allerton/Pelham Gardens    Boro Zone

In [48]:
tzl["LocationID"]

0        1
1        2
2        3
3        4
4        5
      ... 
260    261
261    262
262    263
263    264
264    265
Name: LocationID, Length: 265, dtype: int64

In [49]:
tzl["service_zone"].unique()

array(['EWR', 'Boro Zone', 'Yellow Zone', 'Airports', nan], dtype=object)

In [50]:
tzl.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 265 entries, 0 to 264
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   LocationID    265 non-null    int64 
 1   Borough       265 non-null    object
 2   Zone          264 non-null    object
 3   service_zone  263 non-null    object
dtypes: int64(1), object(3)
memory usage: 8.4+ KB


In [51]:
zonas_manh = tzl["LocationID"][tzl["Borough"] == "Manhattan"]
zonas_manh

3        4
11      12
12      13
23      24
40      41
      ... 
245    246
248    249
260    261
261    262
262    263
Name: LocationID, Length: 69, dtype: int64

Conservamos únicamente las zonas que están en Manhattan.

# Vehicle Fuel Economy Data

Autos de combustión, tipo de carga, tipo de combustible, tipo de barrels

In [52]:
vfed = pd.read_csv("DATA/Vehicle Fuel Economy Data.csv", low_memory=False)

vfed.head(3)

Year Manufacturer           Model  barrels08  barrelsA08  charge240  \
0  2017          BMW        M6 Coupe  17.500588         0.0        0.0   
1  2017          BMW    X3 sDrive28i  12.396250         0.0        0.0   
2  2016     Infiniti  Q50S Red Sport  13.523182         0.0        0.0   

   city08  city08U  cityA08  cityA08U  ...  mfrCode  c240Dscr  charge240b  \
0    15.0     14.0      0.0       0.0  ...      BMX       NaN         0.0   
1    21.0     21.0      0.0       0.0  ...      BMX       NaN         0.0   
2    20.0     19.0      0.0       0.0  ...      NSX       NaN         0.0   

   c240bDscr   createdOn  modifiedOn  startStop  phevCity  phevHwy  phevComb  
0        NaN  04-04-2016  09-09-2016          Y       0.0      0.0       0.0  
1        NaN  04-04-2016  22-11-2016          Y       0.0      0.0       0.0  
2        NaN  04-04-2016  26-09-2016          N       0.0      0.0       0.0  

[3 rows x 82 columns]

In [53]:
vfed.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 46186 entries, 0 to 46185
Data columns (total 82 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Year             46186 non-null  int64  
 1   Manufacturer     46186 non-null  object 
 2   Model            46176 non-null  object 
 3   barrels08        46169 non-null  float64
 4   barrelsA08       46169 non-null  float64
 5   charge240        46169 non-null  float64
 6   city08           46169 non-null  float64
 7   city08U          46169 non-null  float64
 8   cityA08          46169 non-null  float64
 9   cityA08U         46169 non-null  float64
 10  cityCD           46169 non-null  float64
 11  cityE            46169 non-null  float64
 12  cityUF           46169 non-null  float64
 13  co2              46169 non-null  float64
 14  co2A             46169 non-null  float64
 15  co2TailpipeAGpm  46169 non-null  float64
 16  co2TailpipeGpm   46169 non-null  float64
 17  comb08      

In [54]:
vfed.columns

Index(['Year', 'Manufacturer', 'Model', 'barrels08', 'barrelsA08', 'charge240',
       'city08', 'city08U', 'cityA08', 'cityA08U', 'cityCD', 'cityE', 'cityUF',
       'co2', 'co2A', 'co2TailpipeAGpm', 'co2TailpipeGpm', 'comb08', 'comb08U',
       'combA08', 'combA08U', 'combE', 'combinedCD', 'combinedUF', 'cylinders',
       'displ', 'drive', 'engId', 'eng_dscr', 'feScore', 'fuelCost08',
       'fuelCostA08', 'fuelType', 'fuelType1', 'ghgScore', 'ghgScoreA',
       'highway08', 'highway08U', 'highwayA08', 'highwayA08U', 'VClass',
       'highwayCD', 'highwayE', 'highwayUF', 'hlv', 'hpv', 'id', 'lv2', 'lv4',
       'mpgData', 'phevBlended', 'pv2', 'pv4', 'range', 'rangeCity',
       'rangeCityA', 'rangeHwy', 'rangeHwyA', 'trany', 'UCity', 'UCityA',
       'UHighway', 'UHighwayA', 'youSaveSpend', 'guzzler', 'trans_dscr',
       'tCharger', 'sCharger', 'atvType', 'fuelType2', 'rangeA', 'evMotor',
       'mfrCode', 'c240Dscr', 'charge240b', 'c240bDscr', 'createdOn',
       'modifiedOn', 's

In [55]:
vfed[0:1]

Year Manufacturer     Model  barrels08  barrelsA08  charge240  city08  \
0  2017          BMW  M6 Coupe  17.500588         0.0        0.0    15.0   

   city08U  cityA08  cityA08U  ...  mfrCode  c240Dscr  charge240b  c240bDscr  \
0     14.0      0.0       0.0  ...      BMX       NaN         0.0        NaN   

    createdOn  modifiedOn  startStop  phevCity  phevHwy  phevComb  
0  04-04-2016  09-09-2016          Y       0.0      0.0       0.0  

[1 rows x 82 columns]

In [56]:
vfed1 = vfed[['Year', 'Manufacturer', 'Model',
       'mpgData','range', 'rangeCity',
       'rangeCityA', 'rangeHwy', 'rangeHwyA', "fuelType", "fuelType1", 'fuelType2', 'phevCity', 'phevHwy', 'phevComb']]

vfed1

Year Manufacturer                          Model mpgData  range  \
0      2017          BMW                       M6 Coupe       N    0.0   
1      2017          BMW                   X3 sDrive28i       N    0.0   
2      2016     Infiniti                 Q50S Red Sport       N    0.0   
3      2017         MINI  John Cooper Works Convertible       N    0.0   
4      2017          Kia                      Sedona SX       N    0.0   
...     ...          ...                            ...     ...    ...   
46181  1993     Plymouth                Sundance/Duster       N    0.0   
46182  1993      Pontiac                       Grand Am       N    0.0   
46183  1993         Saab                            900     NaN    NaN   
46184  1993       Subaru                         Legacy       N    0.0   
46185  1993       Subaru                     Legacy AWD       Y    0.0   

       rangeCity  rangeCityA  rangeHwy  rangeHwyA fuelType         fuelType1  \
0            0.0         0.0       0.0        0.0  Premium  Premium Gasoline   
1            0.0         0.0       0.0        0.0  Premium  Premium Gasoline   
2            0.0         0.0       0.0        0.0  Premium  Premium Gasoline   
3            0.0         0.0       0.0        0.0  Premium  Premium Gasoline   
4            0.0         0.0       0.0        0.0  Regular  Regular Gasoline   
...          ...         ...       ...        ...      ...               ...   
46181        0.0         0.0       0.0        0.0  Regular  Regular Gasoline   
46182        0.0         0.0       0.0        0.0  Regular  Regular Gasoline   
46183        NaN         NaN       NaN        NaN      NaN               NaN   
46184        0.0         0.0       0.0        0.0  Regular  Regular Gasoline   
46185        0.0         0.0       0.0        0.0  Regular  Regular Gasoline   

      fuelType2  phevCity  phevHwy  phevComb  
0           NaN       0.0      0.0       0.0  
1           NaN       0.0      0.0       0.0  
2           NaN       0.0      0.0       0.0  
3           NaN       0.0      0.0       0.0  
4           NaN       0.0      0.0       0.0  
...         ...       ...      ...       ...  
46181       NaN       0.0      0.0       0.0  
46182       NaN       0.0      0.0       0.0  
46183       NaN       NaN      NaN       NaN  
46184       NaN       0.0      0.0       0.0  
46185       NaN       0.0      0.0       0.0  

[46186 rows x 15 columns]

In [57]:
vfed1["fuelType"].unique()

array(['Premium', 'Regular', 'Diesel', 'Electricity', 'Gasoline or E85',
       'Midgrade', 'Premium and Electricity', 'Premium or E85', nan,
       'Regular Gas and Electricity', 'Regular Gas or Electricity',
       'Premium Gas or Electricity', 'CNG', 'Gasoline or natural gas',
       'Gasoline or propane'], dtype=object)

In [60]:
vfed1.isna().sum()

Year                0
Manufacturer        0
Model              10
mpgData          2007
range            1994
rangeCity        1994
rangeCityA       1994
rangeHwy         1994
rangeHwyA        1994
fuelType         1233
fuelType1        1233
fuelType2       44539
phevCity         2016
phevHwy          2016
phevComb         2016
dtype: int64